fdbclient/NativeAPI.actor.cpp

```diff
@@ -3174,32 +3174,69 @@ ACTOR Future<Void> trackBoundedStorageMetrics(
 	}
 }
 
-ACTOR Future< StorageMetrics > waitStorageMetricsMultipleLocations(
-	vector< pair<KeyRange,Reference<LocationInfo>> > locations,
-	StorageMetrics min,
-	StorageMetrics max,
-	StorageMetrics permittedError)
-{
+ACTOR Future<StorageMetrics> waitStorageMetricsMultipleLocations(
+    vector<pair<KeyRange, Reference<LocationInfo>>> locations, StorageMetrics min, StorageMetrics max,
+    StorageMetrics permittedError, bool largeKeyRange, Database cx) {
 	state int nLocs = locations.size();
-	state vector<Future<StorageMetrics>> fx( nLocs );
+	state vector<Future<StorageMetrics>> fx(largeKeyRange ? nLocs * 2 : nLocs);
 	state StorageMetrics total;
 	state PromiseStream<StorageMetrics> deltas;
 	state vector<Future<Void>> wx( fx.size() );
 	state StorageMetrics halfErrorPerMachine = permittedError * (0.5 / nLocs);
 	state StorageMetrics maxPlus = max + halfErrorPerMachine * (nLocs-1);
 	state StorageMetrics minMinus = min - halfErrorPerMachine * (nLocs-1);
 
-	for(int i=0; i<nLocs; i++) {
-		WaitMetricsRequest req(locations[i].first, StorageMetrics(), StorageMetrics());
-		req.min.bytes = 0;
-		req.max.bytes = -1;
-		fx[i] = loadBalance( locations[i].second, &StorageServerInterface::waitMetrics, req, TaskPriority::DataDistribution );
-	}
-	wait( waitForAll(fx) );
+	state int i = 0;
+	if (largeKeyRange) {
+		// The caller wish to query a huge key-range, could be as large as the whole DB, and thus we do not want to
+		// re-call the API for all locations when the key-range got moved.
+		for (; i < nLocs; i++) {
+			WaitMetricsRequest req(locations[i].first, StorageMetrics(), StorageMetrics());
+			req.min.bytes = 0;
+			req.max.bytes = -1;
+			try {
+				fx[i] = loadBalance(locations[i].second, &StorageServerInterface::waitMetrics, req,
+				                    TaskPriority::DataDistribution);
+				StorageMetrics tmp = wait(fx[i]);
```

Comment:
Instead of wait on each of the request, change to recursive calls to the same function to achieve better parallelism.